feat: Unity Transport + Relay

cdmazom opened the pull request:
This PR adds support for using the Unity Transport package as a transport for MLAPI.

wackoisgod committed:
feat: Initial new UTP Transport

wip still waiting on some packages to be published.

wackoisgod committed:
chore: forgot these files

andrews-unity committed:
Removing registry

wackoisgod committed:
feat: Some UI for authenticating and setup for relay

Relay SDK stuff

wackoisgod committed:
feat: pushing latest stuff

Pushing latest

wackoisgod committed:
chore: fixing define and changing buffer to 1024

wackoisgod committed:
chore: updating package

updating UTP to preview 5 with Relay ping support
fixing issue with endpoint

wackoisgod committed:
Chore: Cleanup

Some cleanup

wackoisgod committed:
chore: file cleanup

cleanup

wackoisgod committed:
fix: Make it so Relay is not a direct depenecy of MLAPI UTP Transport

In order to use Relay for this transport you need to install the package into your project its no longer a dependcy

cdmazom committed:
Re-enable Relay Transport on Sample project

cdmazom committed:
Merge remote-tracking branch 'origin/develop' into feature/initial-relay-utp

fjgal committed:
Update Relay SDK version (#909)

* update to relay sdk v0.0.1-preview.5
* remove workarounds
  - use byte array representation of allocation id
  - connection data truncation is not longer needed
  - set default relay endpoint as production
* updated auth to v0.5.0
  - updated core and authentication versions
  - updated relay version (compatible with latest core)
  - updated authentication calls

unity-cla-assistant commented:
[![CLA assistant check](https://cla-assistant.cds.internal.unity3d.com/pull/badge/not_signed)](https://cla-assistant.cds.internal.unity3d.com/Unity-Technologies/com.unity.multiplayer.mlapi?pullRequest=887) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.cds.internal.unity3d.com/Unity-Technologies/com.unity.multiplayer.mlapi?pullRequest=887) before we can accept your contribution.<br/>**1** out of **2** committers have signed the CLA.<br/><br/>:white_check_mark: andrews-unity<br/>:x: wackoisgod<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.cds.internal.unity3d.com/check/Unity-Technologies/com.unity.multiplayer.mlapi?pullRequest=887) it.</sub>

Jura-Z committed:
Faster pointer and ref based UTP API usage (#912)

fjgal committed:
use allocation id bytes instead of guid

C# Guid convertion to byte array is not compatbile with relay, intead we can use the byte array provided by the relay sdk

andrews-unity committed:
Merge branch 'develop' into feature/initial-relay-utp

wackoisgod committed:
bug: Fixing Packages Version

andrews-unity committed:
bug: Adding support for frag pipeline

andrews-unity committed:
Merge branch 'develop' into feature/initial-relay-utp

andrews-unity committed:
Updating package files

andrews-unity committed:
Exposing the Fragmentation Capacity which was too small for boss room

simon-lemay-unity committed:
Generate Connect events from accepted connections (#969)

MLAPI appears to expect Connect events on servers when new clients
connect (at least this is what's implemented in SIPTransport). UTP
doesn't generate Connect events server-side, so we must generate them
ourselves when accepting connections.

simon-lemay-unity committed:
UTP: Add some tests for the transport wrapper (#976)

* Add some basic UTPTranport tests

* Implement OnDestroy in UTPTransport

The implementation is the same as Shutdown, meaning it only disposes of
the driver (if not already disposed of). The driver (!) for this change
is to avoid leaking resources when tests fail.

* Add connection tests for UTPTransport

* Add basic data transmission tests for UTPTransport

simon-lemay-unity committed:
UTP: Add proper support for channels (#986)

* Correctly handle empty messages in Send

The UTPTransport.Send method would throw a NullReferenceException
exception when passed default(ArraySegment<byte>) as the message data.
Now it simply sends an empty message to its remote host (not a terribly
useful feature, but better than an exception).

* Make it possible to receive messages >1024 bytes

The fragmentation pipeline was set up to handle messages up to 6KB in
length, but the receive buffer was only 1KB long. Messages longer than
that were just ignored (although an error message was logged).

* Add proper support for channels in UTPTransport

The transport now selects an appropriate pipeline according to the
delivery level of the channel, rather than send everything on the
default unreliable pipeline.

thoward-unity committed:
Feature/initial relay utp remove relay sdk calls (#1032)

* DST-528: Extracted Relay SDK calls out of UTP adapter for MLAPI; it now accepts a variable passed in via a setter method. Also updated the sample test project included to illustrate these changes.

* DST-528: Made some changes that address feedback comments; mainly restructured the manner that code external to UTPTransport passes in the requisite relay server data so that external code doesn't need to know about the internal data structures that UTPTransport uses.

* DST-528: Removing minor changes that were erroneously left in when taking other approaches.

thoward-unity committed:
Revert "Feature/initial relay utp remove relay sdk calls (#1032)" (#1033)

This reverts commit ccbf652dc747726eb5f72096d23a6f415b399a16.

thoward-unity committed:
Feature/initial relay utp remove relay sdk calls (#1034)

* DST-528: Extracted Relay SDK calls out of UTP adapter for MLAPI; it now accepts a variable passed in via a setter method. Also updated the sample test project included to illustrate these changes.

* DST-528: Made some changes that address feedback comments; mainly restructured the manner that code external to UTPTransport passes in the requisite relay server data so that external code doesn't need to know about the internal data structures that UTPTransport uses.

* DST-528: Removing minor changes that were erroneously left in when taking other approaches.

* DST-528: Addressing minor code review changes.

* Revert "Revert "Feature/initial relay utp remove relay sdk calls (#1032)" (#1033)"

This reverts commit cd545291db9e6fed5c08474cdc704397e2e271ca.

andrews-unity committed:
Adding some more options exposed and also logging the result of a failed begin/end send

andrews-unity committed:
Merge branch 'develop' into feature/initial-relay-utp

andrews-unity committed:
fix: Fixing broken tests and compile issues

andrews-unity committed:
Fix: Removing any other refs to MLAPI

andrews-unity committed:
feat: rename utp package to align with netcode branding rename.

Cosmin-B committed:
Merge branch 'develop' into feature/initial-relay-utp

Cosmin-B committed:
feat: add a send queue in the UTP adapter so that we are not sending the messages right away, instead we add them to a queue and flush that queue at the end of the frame

Cosmin-B committed:
fix: an edge case where the send queue buffer would become full and we would not clear and send that data is covered, fix some minor bugs where I writing the wrong number of bytes + flushing send queues now happens in Update rather than LateUpdate + more clean-up and added some more comments

Cosmin-B committed:
fix: remove commented code

Cosmin-B committed:
Replace MemoryStream with a NativeArray + DataStreamWriter

Cosmin-B committed:
fix: a minor exception when exiting play mode in editor + clean-up all debug logs and comments

Cosmin-B committed:
Merge branch 'feature/utp-adapter-batch-messages' into feature/initial-relay-utp

Cosmin-B committed:
fix: implement the IEquatable interface for the SendTarget struct to avoid any  allocations when accessing the sendQueue dictionary

Cosmin-B committed:
Merge branch 'develop' into feature/initial-relay-utp

Cosmin-B committed:
Update UTPTransport adapter to latest NGO develop, NetworkChannels are now replaced by QoS delivery hence we no longer need to send them between server and client + update all the UTP Adapter tests and remove the ones that are no longer needed

andrews-unity committed:
Fix: Depending on all public packages

* Depending on all public packages
* Fixed usage of relay SDK (using SDK API vs the other API we were using)
* Updated NGO Package to use pre-release version of collections
* Updated usage of FixedString because of name change in the collection packages
* Removed dependecy on jobs package as its internal

andrews-unity committed:
Fix: Fixing spelling error

andrews-unity committed:
Fix: Fixing tests

Fixing tests and fixing coding standards for Adapter project.

andrews-unity committed:
Merge branch 'develop' into feature/initial-relay-utp

andrews-unity committed:
Fix: Fixing wrong path and package reference.

andrews-unity committed:
Merge branch 'feature/initial-relay-utp' of github.com:Unity-Technologies/com.unity.multiplayer.mlapi into feature/initial-relay-utp

andrews-unity committed:
Fix: Coding standards

andrews-unity committed:
Fix: Fixing coding standards

Cosmin-B commented on testproject/Assets/Scripts/RelayUtility.cs:
```diff
@@ -0,0 +1,58 @@
+
+public class RelayUtility
+{
+    async public static Task<(string ipv4address, ushort port, byte[] allocationIdBytes, byte[] connectionData, byte[] key, string joinCode)> AllocateRelayServerAndGetJoinCode(int maxConnections, string region = null)
```
The region parameter is not used

andrews-unity committed:
Fix: Allow for the ability to provide a new driver/pipeline constructor

andrews-unity committed:
Fix: Fixing a copy pasta error

andrews-unity committed:
test: up the timeout to 150ms

We give a little extra room for the timeout because the test runner may launch a series of tests and thus impact the 50ms time 150ms seems to be reasonable.

andrews-unity committed:
fix: Fixing coding standards

andrews-unity committed:
Fix: Fixing more whitespace issues

andrews-unity committed:
chore: Adding in some tools for simulator stuff

andrews-unity committed:
Rolling back to an older version of UTP to see if tests pass

andrews-unity committed:
Revert "Rolling back to an older version of UTP to see if tests pass"

This reverts commit ea523fd837ee4085de6540590ee1755c11552c60.

andrews-unity committed:
Testing a bug fix :P

andrews-unity committed:
Going back to uNET

andrews-unity committed:
fix: fixing a compile issue

andrews-unity committed:
Fixing compile issue

andrews-unity committed:
Trying to ensure things get cleaned up after every test

andrews-unity committed:
test: changing timeout logic for macOS and coding standard cleanup

andrews-unity committed:
test: again change time

andrews-unity committed:
fix: coding standards

andrews-unity committed:
Merge branch 'develop' into feature/initial-relay-utp

andrews-unity committed:
fix: fix bad merge

andrews-unity committed:
[skip ci] fix: making sure to pass down the region for relay

andrews-unity committed:
[skip ci] removing stuff out of settings files

0xFA11 commented on com.unity.netcode.adapter.utp/CHANGELOG.md:
```diff
@@ -2,4 +2,4 @@
 
 ## [0.0.1-preview.1] - 2020-12-20
-This is the first release of Unity MLAPI Package
+This is the first release of Unity Transport for Netcode for Gameobjects
```
nit:
```suggestion
This is the first release of Unity Transport for Netcode for GameObjects
```

also shouldn't this changelog entry be different? 1.0.0-pre and new date (today?)

andrews-unity replied:
We will fix the changelog in another PR.

0xFA11 commented on com.unity.netcode.adapter.utp/Editor/MultiplayerWindow.cs:
```diff
@@ -0,0 +1,60 @@
+
+namespace Unity.Netcode.Editor
+{
+    public class MultiplayerWindow : EditorWindow
```
hmm, what's this?

Cosmin-B replied:
We added a quick editor for users to be able to use the Simulation Pipeline and tweak various settings about it

andrews-unity replied:
As @Cosmin-B said this was at the request of the samples team, essentially DOTS Netcode has something similar

0xFA11 commented on com.unity.netcode.adapter.utp/Runtime/com.unity.netcode.adapter.utp.asmdef:
```diff
@@ -0,0 +1,20 @@
+{
+    "name": "Unity.Netcode.Adapter.UTP",
+    "rootNamespace": "",
```
nit:
```suggestion
    "rootNamespace": "Unity.Netcode.UTP",
```

0xFA11 commented on testproject/ProjectSettings/ProjectVersion.txt:
```diff
@@ -1,2 +1,2 @@
-m_EditorVersion: 2020.3.12f1
-m_EditorVersionWithRevision: 2020.3.12f1 (b3b2c6512326)
+m_EditorVersion: 2020.3.15f2
+m_EditorVersionWithRevision: 2020.3.15f2 (6cf78cb77498)
```
unnecessary?

andrews-unity replied:
I mean personally I think the project should always be on the latest version of the LTS :P but that is just me. which would be .18 :P I me right now with PR triggers we aren't even testing on 2020.3 which I find also wrong.

0xFA11 replied:
nightly runs on various platforms and versions at least, but yeah 2020.3 baseline for PRs might be more useful, I'm not sure

andrews-unity replied:
Well seeing as how its the min version for the package it should run on it :P I mean for UTP we run on 2020.3/2021.2/trunk across all desktop platforms for PRs because more often then I would like to admit stuff breaks across unity versions. I think we should at least make sure PRs work with the min version.

0xFA11 commented on testproject/ProjectSettings/ProjectSettings.asset:
```diff
@@ -672,8 +680,8 @@ PlayerSettings:
   cloudProjectId: 
   framebufferDepthMemorylessMode: 0
   qualitySettingsNames: []
-  projectName: 
-  organizationId: 
+  projectName: relay-stg
+  organizationId: relay-stg
```
not needed?

andrews-unity replied:
Yes, those values are for the cloud services related stuff since we did add relay testing to the project. So for internal testing that is what they need to be, if public devs want to mess around with things this value will automatically get updated with their services info when they login to their services uDash. So for us yes its required.

andrews-unity committed:
Update ProjectVersion.txt

andrews-unity committed:
Merge branch 'develop' into feature/initial-relay-utp

simon-lemay-unity commented on com.unity.netcode.adapter.utp/Editor/MultiplayerWindow.cs:
```diff
@@ -0,0 +1,60 @@
+        [MenuItem("Netcode/NGO Tools")]
+        public static void ShowWindow()
+        {
+            GetWindow<MultiplayerWindow>(false, "NGO Tools Tools", true);
```
Minor, but is the duplicated "Tools" here on purpose?

andrews-unity replied:
no will fix :P

simon-lemay-unity commented on com.unity.netcode.adapter.utp/Runtime/UTPTransport.cs:
```diff
@@ -0,0 +1,835 @@
+using System.Linq;
+using UnityEngine;
+
+using UTPNetworkEvent = Unity.Netcode.NetworkEvent;
```
Is there a reason for renaming the NGO `NetworkEvent` as `UTPNetworkEvent`? I understand why we're renaming in the first place (to avoid a namespace conflict), but I would expect the `NetworkEvent` from UTP to be named `UTPNetworkEvent`, not the NGO one.

andrews-unity replied:
That was rider renaming :P the main issue because conflicting Types since UTP and NGO both have a NetworkEvent type

simon-lemay-unity commented on com.unity.netcode.adapter.utp/Runtime/UTPTransport.cs:
```diff
@@ -0,0 +1,835 @@
+
+            InitDriver();
+
+            if (m_Driver.Bind(NetworkEndPoint.AnyIpv4) != 0)
```
Is IPv6 supported by the adapter? If so, might want to choose between this and `AnyIpv6` depending on the address family of `serverEndpoint`.

Alternatively, I don't think there's any need to bind on the client side. Could remove this entirely.

andrews-unity replied:
We do support IPv6 in the transport but I don't think we support it currently in the adapter. 

As for binding the protocol can bind which in this case we use for Relay

andrews-unity replied:
I will add a ticket to support IPV6 to the transport API in general.

simon-lemay-unity commented on com.unity.netcode.adapter.utp/Runtime/UTPTransport.cs:
```diff
@@ -0,0 +1,835 @@
+                    ;
+                }
+
+                FlushAllSendQueues();
```
Shouldn't we do that before the driver update? This way the driver's send job will be able to immediately send it all on the network. Otherwise we wait until the next `Update` to send anything.

andrews-unity replied:
cc @Cosmin-B as he wrote this code

Cosmin-B replied:
Moved it before the driver update, thanks!

andrews-unity committed:
Addressing some PR feedback

Cosmin-B committed:
fix: move FlushSendQueues call before we do drive update so that we don't delay the queued messages by another tick

andrews-unity committed:
Merge branch 'develop' into feature/initial-relay-utp

0xFA11 commented on com.unity.netcode.adapter.utp/Runtime/UTPTransport.cs:
```diff
@@ -0,0 +1,836 @@
+            Connected,
+        }
+
+        public const int MaximumMessageLength = 6 * 1024;
```
where this number is coming from?

andrews-unity replied:
Boss Room :( It is still configurable in the serialized field.

0xFA11 commented on com.unity.netcode.adapter.utp/Runtime/UTPTransport.cs:
```diff
@@ -0,0 +1,836 @@
+using System.Linq;
+using UnityEngine;
+
+using NGONetworkEvent = Unity.Netcode.NetworkEvent;
+using UTPNetworkEvent = Unity.Networking.Transport.NetworkEvent;
```
```suggestion
using NetcodeNetworkEvent = Unity.Netcode.NetworkEvent;
using TransportNetworkEvent = Unity.Networking.Transport.NetworkEvent;
```

0xFA11 commented on com.unity.netcode.adapter.utp/Runtime/UTPTransport.cs:
```diff
@@ -0,0 +1,836 @@
+        void CreateDriver(UTPTransport transport, out NetworkDriver driver, out NetworkPipeline unreliableSequencedPipeline, out NetworkPipeline reliableSequencedPipeline, out NetworkPipeline reliableSequencedFragmentedPipeline);
+    }
+
+    public class UTPTransport : NetworkTransport, INetworkStreamDriverConstructor
```
```suggestion
    public class UnityTransport : NetworkTransport, INetworkStreamDriverConstructor
```

0xFA11 commented on com.unity.netcode.adapter.utp/Runtime/UTPTransport.cs:
```diff
@@ -0,0 +1,836 @@
+        [SerializeField] private int m_ReciveQueueSize = 128;
+        [SerializeField] private int m_SendQueueSize = 128;
+
+        [Tooltip("The maximum size of the send queue for batching NGO events")]
```
```suggestion
        [Tooltip("The maximum size of the send queue for batching Netcode events")]
```

0xFA11 commented on com.unity.netcode.adapter.utp/Runtime/UTPTransport.cs:
```diff
@@ -0,0 +1,836 @@
+                    return m_ReliableSequencedFragmentedPipeline;
+
+                default:
+                    Debug.LogError($"Unknown NetworkDelivery value: {delivery}");
```
nit:
```suggestion
                    Debug.LogError($"Unknown {nameof(NetworkDelivery)} value: {delivery}");
```

0xFA11 commented on com.unity.netcode.adapter.utp/Tests/Editor/UTPTransportTests.cs:
```diff
@@ -0,0 +1,92 @@
+
+namespace Unity.Netcode.UTP.EditorTests
+{
+    public class UTPTransportTests
```
```suggestion
    public class UnityTransportTests
```

andrews-unity committed:
PR Feedback

0xFA11 reviewed (approved):
some naming nits but apart from that, looks good to me!
(UTP -> UnityTransport & NGO -> Netcode)

andrews-unity committed:
Fixing rename issues

andrews-unity committed:
Merge branch 'develop' into feature/initial-relay-utp
